# Part 1 : Python Basics.
(Followed DSMP CampusX Lectures)

## Why Python?

**1. Easy to Learn & Read**
- Simple, clean syntax (like plain English).
- Great for beginners, yet powerful for advanced projects.

**2. Batteries Included**
- Comes with a rich standard library (no extra tools needed for basic tasks).
- Examples: File handling, web scraping, math operations, etc.

**3. Versatile (General Purpose)**
- Used in multiple fields: like Web Devlopment, Data Science, Automation, Game Development etc. 

**4.  Huge Community & Libraries**
- Thousands of free libraries for almost any task. (eg. Numpy, Tensorflow, Django)
- Strong community support → Easy troubleshooting, tutorials, and updates.

## Why python for data science?
1. Easy to learn
2. Proximity  with Maths
3. Community


> **Note:** 
1. Python is case-sensitive language. **print** and **Print** are two different thing
2. Python is a **interpreted language** not a ~~compiled language~~, it is translated line by line.


---

## Python Syntax:


### 1. Python Output:

In [1]:
print('Hello World')

Hello World


In [3]:
print('Hello World', 1, 3.14 , True)

Hello World 1 3.14 True


> **Note:** In Python, when we pass multiple elements to the ``print()`` function, all elements are printed separated by a space by default.
However, we can change this behavior using the ``sep`` (separator) parameter of ``print()``.

In [1]:
print('Hello World', 1, 3.14 , True , sep = '/')

Hello World/1/3.14/True


> **Note:** By default in Python, after the ``print()`` function executes, the output moves to a new line.
We can change this behavior using the ``end`` parameter of print().
By default ``end = '\n'``;

In [3]:
# By default end='\n'
print('Shivam')
print('Sinha')

Shivam
Sinha


In [ ]:
# changing behaviour of 'end' parameter
print('Shivam' , end = "-")
print('Sinha')

Shivam-Sinha


---

### 2. Data Types

In [ ]:
# Integer
print(8)

# Python can handle a large integer values
# About 1*10^308
print(1e308)

8
1e+308


In [9]:
print(1e309)

inf


infinite because it can handle only upto 10^308

In [11]:
# Decimal / Float

print(8.55)
print(1.7e308)

8.55
1.7e+308


In [12]:
## Boolean

print(True)
print(False)

True
False


In [16]:
# Text / String

print("Hello World")

Hello World


In [15]:
# Complex

print(5+6j)

(5+6j)


In [17]:
# List (Array in Java or C)

print([1,2,3,4,5])

[1, 2, 3, 4, 5]


In [19]:
# Tuple

print((1,2,3,4,5))

(1, 2, 3, 4, 5)


In [20]:
# Sets

print({1,2,3,4,5,5})

{1, 2, 3, 4, 5}


In [21]:
# Dictionary - Key value pairs

print({'Name':'Shivam' , 'Gender' : 'Male' , 'Weight' : 60})

{'Name': 'Shivam', 'Gender': 'Male', 'Weight': 60}


In [22]:
# Type

type(3)

int

In [26]:
type(3.5)


float

In [27]:
type("Hello")

str

In [28]:
type(3.14)

float

In [29]:
type(6+6j)

complex

In [30]:
type([1,2,3])

list

--- 

### 3. Variables
- Variables are basically containers, used to store data for later use.
- In python, we don't need to explicitly mention the data type.
- Python is dynamically typed - it automatically detects the data type based on the assigned value

In [37]:
name = "Shivam"
print(name)
print(type(name))

a = 5
b = 6 
print( a+ b )

Shivam
<class 'str'>
11


---
### Dynamic Typing vs Static Typing

**Dynamic Typing**: When we **don't need to specify the data type** explicitly.  
Python supports dynamic typing.

``
a = 5  # Python automatically understands that 'a' is an integer
``

**Static Typing** : We must specify the data type explicitly. C,C++, Java supports static typing
``
int a = 5
``

---

### Dynamic binding vs Static binding

**Dynamic binding**: Once a variable is declared, its **data type can changed** at anytime during in the program. 
Python supports dynamic binding.


```py
a = 5 
a = 'shivam'   # No error in python
```

**Static binding** : Once a variable is declared, its the data type remains fixedthroughout the program. C,C++, Java supports static binding

```java
int a = 5

String a = "Shivam";  //Error in Java and C++ as 'a' is already declared as integer.
```

### More ways to create multiple variables

In [40]:
a,b,c = 1,2,3
print(a,b,c)

1 2 3


In [42]:
a=b=c = 5
print(a,b,c)

5 5 5


> **Note:** compilation is basically a process where we convert our high level code into binary language (machine language)

--- 

### 4. Keywords & Identifiers

**Keywords**
- Keywords are reserved words in python that have special meaning 
- They cannot be used as identifiers (variable names, functions names, etc.)
- They are reserved by the **python interpreter** and used by them to understand the structure and logic of the program better.
- In python there are 35 keywords

| `False`    | `None`    | `True`  | `and`   | `as`     |
| ---------- | --------- | ------- | ------- | -------- |
| `assert`   | `async`   | `await` | `break` | `class`  |
| `continue` | `def`     | `del`   | `elif`  | `else`   |
| `except`   | `finally` | `for`   | `from`  | `global` |
| `if`       | `import`  | `in`    | `is`    | `lambda` |
| `nonlocal` | `not`     | `or`    | `pass`  | `raise`  |
| `return`   | `try`     | `while` | `with`  | `yield`  |


**Identifiers** 
- Identifiers are the names you assign to elements in your Python code — such as: Vaiables, Functions, Classes, Modules, Objects

**Rules for writing Identifiers in python:**
1. Can contain letters (A–Z, a–z), digits (0–9), and underscores (_)
2. Cannot start with a digit
3. Cannot be a keyword (like if, while, class, etc.)
4. Are case-sensitive (Name and name are different)
5. No special characters allowed (@, $, #, etc.)

---

### 5. User Input

In [47]:
s = input("Enter Input")
print(s)

Shivam


### WAP in python to add 2 number

In [50]:
num1 = input("Enter first number")
num2 = input("Enter second number")
sum = num1 + num2
print("The sum of: ", num1 , "and" , num2, "is" , sum)

The sum of:  5 and 15 is 515


> **Note:** Behaviour of input() function in python
- In Python, the `input()` function **always returns data as a string**, even if the user enters a number.
- If you try to perform arithmetic directly on `input()` values, Python will **concatenate the strings** instead of adding the numbers.
- So the correct approach is, always convert input() to int or float when performing mathematical operations.


In [51]:
num1 = int(input("Enter first number"))
num2 = int(input("Enter second number"))
sum = num1 + num2
print("The sum of: ", num1 , "and" , num2, "is" , sum)

The sum of:  5 and 15 is 20


---
### 6. Type Conversion 
- Type conversion refers to changing the data type of a value from one type to another, such as converting an **int** to a **float**, or a **string** to an **int**.
- Python supports two kinds of type conversion

    1. **Implicit Type Conversion**: Automatically done by python when needed
    
    2. **Explicit Type Conversion** : Also known as type casting, where the programmer manually converts the datatype using built-in functions like (int(), float(), str(), list() etc...)


In [3]:
print(5+5.6)
print(type(5), type(5.6))

10.6
<class 'int'> <class 'float'>


In [7]:
# str -> int
a = int('4')
print(type(a))

# int -> str
print(str(5))

<class 'int'>
5


---
### 7. Different types of literals 

In [8]:
a = 0b1010  # Binary literals
b = 100     # Decimal Literals
c = 0o310   # Octal Literals
d = 0x12c   # Hexadecimnal Literals
print(a,b,c,d)

10 100 200 300


In [9]:
# Float litrals
f1 = 10.5
f2 = 1.5e2  # 1.5 * 10^2
f3 = 1.5e-3 # 1.5 * 10^-2
print(f1,f2,f3)

10.5 150.0 0.0015


In [17]:
# String Literals
str1 = 'This is a string'
str2 = "This is a string"
char = "A"
multiline_str = """This is a multiline string"""
unicode = u"\U0001f600\U0001f606\U0001F923"
raw_str = r"raw \n string"
not_raw = "notraw \n string"

print(str1)
print(str2)
print(char)
print(multiline_str)
print(unicode)
print(raw_str)
print(not_raw)


This is a string
This is a string
A
This is a multiline string
😀😆🤣
raw \n string
notraw 
 string


> Note: 
- The **r** before the string means it's a raw string literal.
- In a raw string, backslashes are treated as literal characters, not as escape characters.

In [18]:
# Boolean Literals
a = True + 4
b = False + 10

print("a: ", a)
print("b: ", b)

a:  5
b:  10


> **Note:** This is becuase python treats boolean as number. The value of ``True is 1`` and ``False is 0`` 

### NONE in python

- In python, we can not just declare a variablewithout assigning a value.



In [19]:
K
a = 5
print("Execute")

NameError: name 'K' is not defined

- To handle this, we use ``None`` as a placeholder.

In [20]:
k = None
a = 5
print("Execute")

Execute


---
### 8. Operators
Operators are symbols that perform operations on values/variables.

**1. Arithmetic : used for basic maths**

| Operator | Meaning             | Example       |
| -------- | ------------------- | ------------- |
| `+`      | Addition            | `3 + 2 = 5`   |
| `-`      | Subtraction         | `5 - 1 = 4`   |
| `*`      | Multiplication      | `2 * 3 = 6`   |
| `/`      | Division (float)    | `5 / 2 = 2.5` |
| `//`     | Floor division      | `5 // 2 = 2`  |
| `%`      | Modulus (remainder) | `5 % 2 = 1`   |
| `**`     | Exponentiation      | `2 ** 3 = 8`  |

**2. Relational : Used to compare values (returns True or False).**

| Operator | Meaning          | Example           |
| -------- | ---------------- | ----------------- |
| `==`     | Equal to         | `5 == 5` → `True` |
| `!=`     | Not equal        | `3 != 2` → `True` |
| `>`      | Greater than     | `5 > 2` → `True`  |
| `<`      | Less than        | `2 < 5` → `True`  |
| `>=`     | Greater or equal | `3 >= 3` → `True` |
| `<=`     | Less or equal    | `2 <= 3` → `True` |

**3. Logical : Used to combine multiple conditions.**

| Operator | Meaning           | Example           |
| -------- | ----------------- | ----------------- |
| `and`    | Both true         | `x > 2 and x < 5` |
| `or`     | At least one true | `x < 2 or x > 10` |
| `not`    | Negation          | `not(x > 5)`      |

**4. Bitwise: Works on binary representatiobn of integers(0s and 1s)**

| Operator | Name           | Example         | Binary Meaning            |
| -------- | -------------- | --------------- | ------------------------- |
| `&`      | AND            | `5 & 3` → `1`   | `0101 & 0011 = 0001`      |
| `^`      | XOR            | `5 ^ 3` → `6`   | `0101 ^ 0011 = 0110`      |
| `~`      | NOT (1's comp) | `~5` → `-6`     | `~0101 = 1010` (inverted) |
| `<<`     | Left Shift     | `5 << 1` → `10` | `0101 << 1 = 1010`        |
| `>>`     | Right Shift    | `5 >> 1` → `2`  | `0101 >> 1 = 0010`        |


**5. Assignment : Used to assign values.**

| Operator | Meaning             | Example                |
| -------- | ------------------- | ---------------------- |
| `=`      | Assign              | `x = 5`                |
| `+=`     | Add and assign      | `x += 2` → `x = x + 2` |
| `-=`     | Subtract and assign | `x -= 1`               |
| `*=`     | Multiply and assign | `x *= 3`               |
| `/=`     | Divide and assign   | `x /= 2`               |

**6. Membership : Used to check if a value is in a sequence (like list, string, etc.)**

| Operator | Meaning       | Example                     |
| -------- | ------------- | --------------------------- |
| `in`     | Exists        | `'a' in 'cat'` → `True`     |
| `not in` | Doesn’t exist | `'x' not in 'cat'` → `True` |



---
### WAP to print sum of digits of a three digit number. 
**Eg.** ``num = 456`` ``Output = 4+5+6 = 15``

In [1]:
num = int(input("Enter a 3 digit number"))

a = num%10
num = num//10

b = num%10
num = num//10

c = num%10


sum = a + b + c

print("Sum: ", sum)

Sum:  15


--- 
### 9. If-Else Statement

### WAP a basic login athentication program
- Let email -> shivamsinha@gmail.com
- Let password -> admin

In [8]:
email = input("Enter Email")
password = input("Enter Password")

if email == "shivamsinha@gmail.com" and password == "admin" :
    print("Login Successfull!")

else:
    print("Invalid Credentials!!!")


Login Successfull!


> Note: Unlike Java and C++, Python doesn't use brackets or semicolons. For ``if...else`` statements, we use indentation ``(with a colon :)`` to indicate what is inside the ``if block`` and what is outside of it.

> Indentation refers to the spaces or tabs added at the beginning of a line of code to show which block it belongs to. In Python, indentation is not optional — it is part of the syntax.

### WAP to find min of 3 numbers


In [5]:
a = int(input("Enter first number"))
b = int(input("Enter second number"))
c = int(input("Enter third number"))

if(a<b and a<c):
    print("Min of ", a, b, c, " is ", a)
elif(b<c):
    print("Min of ", a, b, c, " is ", b)
else:
    print("Min of ", a, b, c, " is ", c)

Min of  10 20 30  is  10


---
# 10. Module in Pyton

- In Java and C++, we use header files to include reusable code. In Python, we use modules for the same purpose.

- A module is a Python file that contains functions, variables, or classes. We can import a module into our code and use its contents.

- Some examples of built-in Python modules are:
1. math
2. keyword
3. datetime
4. random 

In [6]:
#math
import math

print(math.sqrt(196))
print(math.floor(6.8))
print(math.factorial(5))

14.0
6
120


In [15]:
#keyword
import keyword

print(keyword.kwlist)
print(keyword.iskeyword('None'))

['False', 'None', 'True', 'and', 'as', 'assert', 'async', 'await', 'break', 'class', 'continue', 'def', 'del', 'elif', 'else', 'except', 'finally', 'for', 'from', 'global', 'if', 'import', 'in', 'is', 'lambda', 'nonlocal', 'not', 'or', 'pass', 'raise', 'return', 'try', 'while', 'with', 'yield']
True


In [20]:
#random
import random

random.randint(1,100)

59

In [24]:
#datetime
import datetime

print(datetime.datetime.now())

2025-07-25 17:13:23.312135


> To check all moudles in python

In [25]:
help('modules')


Please wait a moment while I gather a list of all available modules...

test_sqlite3: testing with SQLite version 3.49.1
IPython             _weakrefset         locale              symtable
__future__          _win32sysloader     logging             sys
__hello__           _win32verstamp_pywin32ctypes lzma                sysconfig
__phello__          _winapi             mailbox             tabnanny
_abc                _winxptheme         marshal             tarfile
_aix_support        _wmi                math                tempfile
_android_support    _zoneinfo           matplotlib_inline   test
_apple_support      abc                 mimetypes           textwrap
_ast                adodbapi            mmap                this
_asyncio            afxres              mmapfile            threading
_bisect             antigravity         mmsystem            time
_blake2             argparse            modulefinder        timeit
_bz2                array               msvcrt             

---
# 11. Loops in Python
- For Loop
- While Loop

## While Loop

In [29]:
i = 1
num = 2

while i<=10:
    print(num , "*", i, "=", num*i)
    i+=1

2 * 1 = 2
2 * 2 = 4
2 * 3 = 6
2 * 4 = 8
2 * 5 = 10
2 * 6 = 12
2 * 7 = 14
2 * 8 = 16
2 * 9 = 18
2 * 10 = 20


## While Loop with else

In [ ]:
i = 1

while(i<5):
    print(i)
    i+=1

else:
    print("Loop Complete")

1
2
3
4
Loop Complete


## For Loop

In [38]:
for i in range (1,11):
    print(i, end=' ')

1 2 3 4 5 6 7 8 9 10 

``range(start, stop)``

In [43]:
for i in 1,3,5,6,7,8:
    print(i)

1
3
5
6
7
8


In [45]:
for i in range(1,20,3): #3 is step size
    print(i)

1
4
7
10
13
16
19


``range(start, stop, step)``

In [47]:
for i in range(10,0,-1): #reverse
    print(i)

10
9
8
7
6
5
4
3
2
1


In [49]:
for i in range(10,0):
    print(i)

#Nothing will print

In [52]:
for i in range(10,0,-2): # Reverse but skip every next element
    print(i)

10
8
6
4
2


> **Note:** Python for loop is so powerful that it can iterate all other data types too. For example

In [ ]:
# for loop on string

for i in 'Delhi':
    print(i)

D
e
l
h
i


In [ ]:
# for loop on List

for i in [1,2,3,4,5]:
    print(i)

1
2
3
4
5


# WAP : In a town of 10,000 people, the population is increasing at a rate of 10% each year. Display population at end of each year (for next 10 years)

In [61]:
population = 1000

for i in range (1,11):
    print(int(population))
    population*=1.1

1000
1100
1210
1331
1464
1610
1771
1948
2143
2357


# WAP : Sequence Sum - (1/1! + 2/2! + 3/3! + 4/4! + ---- + n/n!)

In [66]:
n = int(input("Enter: "))
result = 0
fact = 1

for i in range (1,n+1):
    fact = fact * n
    result = result + n/fact

print("You Entered : ", n)
print("Result: ", result)

You Entered :  5
Result:  1.2496


# Pattern 1
```py
*
* *
* * *
* * * *
* * * * *
```

In [71]:
for i in range(1,6):
    for j in range(1,i+1):
        print('*', end=' ')
    print()

* 
* * 
* * * 
* * * * 
* * * * * 


# Pattern 2
```py
1
1 2 1
1 2 3 2 1 
1 2 3 4 3 2 1
```

In [78]:
for i in range(1,5):
    for j in range (1,i+1):
        print(j, end=" ")
    for k in range (i-1, 0, -1):
        print(k, end=" ")
    print()

1 
1 2 1 
1 2 3 2 1 
1 2 3 4 3 2 1 


--- 
# 12. Loop Control Statement
1. Break
2. Continue
3. Pass

### Break

In [ ]:
for i in range(1,10):
    print(i)
    if i==5:
        break

1
2
3
4
5


# WAP to find all prime number between 1 and 100


In [ ]:
a = 2       # 1 is neither prime nor composite    
b = 100

for i in range(a,b+1):
    for j in range(2,i):
        if i%j == 0:
            break
    else:
        print(i, end=" ")

2 3 5 7 11 13 17 19 23 29 31 37 41 43 47 53 59 61 67 71 73 79 83 89 97 

### Continue

In [92]:
for i in range(1,10):
    if(i==5):
        continue
    print(i)

1
2
3
4
6
7
8
9


### Pass 
Sometimes, we want to write a loop or a function as a placeholder for future logic — but Python doesn’t allow empty blocks.

In [95]:
for i in range(1,10):

_IncompleteInputError: incomplete input (4054113036.py, line 1)

Python expects at least one statement inside a block.

To resolve this we can use **Pass Statement**

- Pass statement does nothing 

In [96]:
for i in range(1,10):
    pass

---
# **Strings**
- Strings are sequence of characters
- In python, strings are a sequence of Unicode Characters

## 1. Creating Strings

In [ ]:
a = 'Hello'
b = "Hello"
c = '''Hello'''
d = """Hello"""
e = str('Hello')
f = str(a)

print(a,b,c,d,e,f, sep = '\n')

Hello
Hello
Hello
Hello
Hello
Hello


## 2. Accessing substring form a string
- **Positive Indexing** - Indexed from ``Left(0) to Right(size-1)``


In [111]:
str = "String"

for i in range(0,len(str)):
    print( str[i], "=", i, "Index ")


S = 0 Index 
t = 1 Index 
r = 2 Index 
i = 3 Index 
n = 4 Index 
g = 5 Index 


- **Negative Indexing** - Indexed from ``Right(-1) to Left(-size)``

<table border="1">
  <tr>
    <td><strong>Positive Indexing</strong></td>
    <td>0</td>
    <td>1</td>
    <td>2</td>
    <td>3</td>
    <td>4</td>
    <td>5</td>
  </tr>
  <tr>
    <td><strong>Character / String</strong></td>
    <td><b>S</b></td>
    <td><b>t</b></td>
    <td><b>r</b></td>
    <td><b>i</b></td>
    <td><b>n</b></td>
    <td><b>g</b></td>
  </tr>
  <tr>
    <td><strong>Negative Indexing</strong></td>
    <td>-6</td>
    <td>-5</td>
    <td>-4</td>
    <td>-3</td>
    <td>-2</td>
    <td>-1</td>
  </tr>
</table>




## 3. String Slicing

In [138]:
str = "Shivam Sinha"
print(str[0:7])

Shivam 


In [139]:
print(str[0:7:2])

Sia 


In [142]:
print(str[6:0:-1])

 mavih


In [143]:
print(str[6::-1])

 mavihS


In [134]:
#reverse the string
print(str[::-1])

ahniS mavihS


In [135]:
print(str[-5:])

Sinha


In [136]:
print(str[-1:-6:-1])

ahniS


## 4. Editing in String
> **Note:** Strings in python are **immutable**. i.e. we cannot change the content of a string or reassign characters after assignment.

In [1]:
str = "shivam"
str[0] = "S"

TypeError: 'str' object does not support item assignment

> Note: Editing of String is not allowed in Python

## 5. Deleting in String
use ``del`` keyword to delete a string

In [ ]:
st = "shivam"
del st
print(st)  #st string is deleted

NameError: name 'st' is not defined

> Successfully Deleted (whole string)

In [ ]:
str = "Shivam Sinha"
del str[-1:-5:2]
print(str)

TypeError: 'str' object does not support item deletion

> **Note:** This code doesn’t work because it attempts to delete a ``portion of a string``. Since ``strings are immutable`` in Python, modifying their content (such as deleting specific characters) is not allowed. However, we can delete the entire string object using the ``del`` statement

> **Note:** Partial deletion of string is not allowerd. Have to delete the whole string.

## 6. Operations on String

### *Arithmetic Operations*
- ``+`` - String Concatenation
- ``*``

In [8]:
# + operator

print('delhi' + ' mumbai')

delhi mumbai


In [10]:
# * operator

print ('a ' * 10)

a a a a a a a a a a 


### *Relationship Operator*

In [11]:
'delhi' == 'mumbai'

False

In [12]:
'delhi' != 'mumbai'

True

In [13]:
'mumbai' > 'pune'

False

> **Note:**  We are not comparing on the basis of no. of string but comparing **Lexiographically** means according to the ASCII value, similar to how the words arranged in a dictionary. Words that come earlier in the dictionary (i.e., with lower ASCII values) are considered smaller.

For example, ``'m' in 'mumbai'`` comes before ``'p' in 'pune'``, so the assumption that ``'mumbai' > 'pune'`` is ``false.``

In [14]:
'Pune' > 'pune'

False

> ASCII value of ``P`` (Capital P) is smaller than ``p`` (Small p)
- ``P``  -> 80   
- ``p``  -> 112

| Dec | Char      | Dec | Char | Dec | Char |
| --- | --------- | --- | ---- | --- | ---- |
| 32  | ␣ (space) | 64  | @    | 96  | \`   |
| 33  | !         | 65  | A    | 97  | a    |
| 34  | "         | 66  | B    | 98  | b    |
| 35  | #         | 67  | C    | 99  | c    |
| 36  | \$        | 68  | D    | 100 | d    |
| 37  | %         | 69  | E    | 101 | e    |
| 38  | &         | 70  | F    | 102 | f    |
| 39  | '         | 71  | G    | 103 | g    |
| 40  | (         | 72  | H    | 104 | h    |
| 41  | )         | 73  | I    | 105 | i    |
| 42  | \*        | 74  | J    | 106 | j    |
| 43  | +         | 75  | K    | 107 | k    |
| 44  | ,         | 76  | L    | 108 | l    |
| 45  | -         | 77  | M    | 109 | m    |
| 46  | .         | 78  | N    | 110 | n    |
| 47  | /         | 79  | O    | 111 | o    |
| 48  | 0         | 80  | P    | 112 | p    |
| 49  | 1         | 81  | Q    | 113 | q    |
| 50  | 2         | 82  | R    | 114 | r    |
| 51  | 3         | 83  | S    | 115 | s    |
| 52  | 4         | 84  | T    | 116 | t    |
| 53  | 5         | 85  | U    | 117 | u    |
| 54  | 6         | 86  | V    | 118 | v    |
| 55  | 7         | 87  | W    | 119 | w    |
| 56  | 8         | 88  | X    | 120 | x    |
| 57  | 9         | 89  | Y    | 121 | y    |
| 58  | :         | 90  | Z    | 122 | z    |
| 59  | ;         | 91  | \[   | 123 | {    |
| 60  | <         | 92  | \\   | 124 | \|   |
| 61  | =         | 93  | ]    | 125 | }    |
| 62  | >         | 94  | ^    | 126 | \~   |
| 63  | ?         | 95  | \_   | 127 | DEL  |


In [22]:
print(chr(80))
print(chr(112))

P
p


### *Logical Operator*
> Python treats strings in logical operations (`and`, `or`, `not`) based on their **true or false** values:

* An **empty string (`''`) is considered `False`**
* A **non-empty string (e.g., `'hello'`) is considered `True`**

> `''` (empty string) → `False`
> `'abc'` (non-empty string) → `True`


> `and` returns the **first false** operand, or the **last operand** if all are true.
> `or` returns the **first true** operand.



In [23]:
'Shivam' and 'Sinha'

'Sinha'

In [24]:
'Shivam' or 'Sinha'

'Shivam'

In [25]:
'' and 'Sinha'

''

In [26]:
'' or 'Shivam'

'Shivam'

In [27]:
'' or ''

''

In [28]:
not 'Shivam'

False

### *Membership Operator*

In [35]:
'D' in "Delhi"

True

In [36]:
'D' not in "delhi"

True

## 7. Loop in Python

In [30]:
for i in 'Shivam':
    print(i)

S
h
i
v
a
m


In [34]:
for i in 'Shivam':
    print("Sinha")

Sinha
Sinha
Sinha
Sinha
Sinha
Sinha


---
## 8. String functions
### Common Functions 
> Works on every datatypes (String, Tuple, List, Set, Dictionary)
- ``len()``
- ``max()``
- ``min()``
- ``sorted()``

In [37]:
len("Shivam Sinha")

12

In [38]:
max("Shivam Sinha")

'v'

> Max and Min based on ASCII value

In [41]:
min("Shivam Sinha")

' '

In [42]:
sorted("Shivam Sinha")

[' ', 'S', 'S', 'a', 'a', 'h', 'h', 'i', 'i', 'm', 'n', 'v']

> **Note:** ``sorted(string)`` returns a list of characters from the string, sorted in ascending (lexicographic) order.

> **Note:** It does not modify the original string, because strings in Python are immutable.

In [46]:
sorted("Shivam Sinha", reverse=True)

['v', 'n', 'm', 'i', 'i', 'h', 'h', 'a', 'a', 'S', 'S', ' ']

### ``capitalize() / title() / upper() / lower() / swapcase()``

In [48]:
str = "Shivam Sinha"
str.capitalize()

'Shivam sinha'

> ``capitalize()`` method returns a new string with the first character capitalized and the rest of the string converted to lowercase.
It does not modify the original string (since strings are immutable in Python).

In [50]:
str.title()

'Shivam Sinha'

> ``The title()`` method returns a new string where the first letter of each word is capitalized, and all other letters are in lowercase.

In [53]:
str.upper()

'SHIVAM SINHA'

In [49]:
str.lower()

'shivam sinha'

In [52]:
str.swapcase()

'sHIVAM sINHA'

### ``count() / find() / index()``

In [55]:
str.count('S')

2

In [56]:
str.find('S')

0

In [58]:
str.find('Sin')

7

In [59]:
str.find('x')

-1

> ``find()`` returns the index of the first occurrence of a specified substring or character in a string. Returns `-1` if **not found**

In [60]:
str.index('S')

0

In [61]:
str.index('Sin')

7

In [62]:
str.index('x')

ValueError: substring not found

> **Note:** Both ``.find()`` and ``.index()`` return the index of the first occurrence of a substring.
However, ``.find()`` returns ``-1`` if not found, while ``.index()`` raises a ``ValueError.``
Use ``.find()`` when you're unsure whether the substring exists.

### ``endswithg() / startswith()``

In [63]:
str.endswith('nha')

True

In [64]:
str.endswith('vam')

False

In [66]:
str.startswith('Shi')

True

In [67]:
str.startswith('shi')

False

### ``Format``

In [69]:
name = "Shivam"
gender = "Male"

"Hi, My name is {} and I am a {} ".format(name,gender)

'Hi, My name is Shivam and I am a Male '

In [71]:
name = "Shivam"
gender = "Male"

"Hi, My name is {} and I am a {} ".format(gender,name)

'Hi, My name is Male and I am a Shivam '

In [70]:
name = "Shivam"
gender = "Male"

"Hi, My name is {1} and I am a {0} ".format(gender,name)

'Hi, My name is Shivam and I am a Male '

### ``isalnum()``
Check if the string is Alpha-Numeric (Alphabet and Number only) or not

In [72]:
'Shivam1234'.isalnum()

True

In [73]:
'Shivam@1234'.isalnum()

False

In [74]:
'Shivam Sinha'.isalnum()

False

### ``isalpha()``
Check if string contains only alphabet or not

In [75]:
'Shivam'.isalpha()

True

In [76]:
'Shivam Sinha'.isalpha()

False

### ``isdigit()``

In [77]:
'1234'.isdigit()

True

### ``isidentifier()``

In [81]:
'1num'.isidentifier()

False

In [83]:
'num1'.isidentifier()

True

In [84]:
'first_name'.isidentifier()

True

In [85]:
'first-name'.isidentifier()

False

### ``split() / join()``

In [86]:
"hi, my name is Shivam Sinha".split()

['hi,', 'my', 'name', 'is', 'Shivam', 'Sinha']

In [89]:
"hi, my name is Shivam Sinha".split('i')

['h', ', my name ', 's Sh', 'vam S', 'nha']

> **Note:** ``Split()`` method breaks a string into a list of substrings based on a separator (default is whitespace), so its output is a ``list``, not a string.

In [91]:
word = ['Shivam', 'Sinha']

" ".join(word)

'Shivam Sinha'

> **Note:** join() method used to combine element of an iterable ``(like a list or a tuple) into a single string``, with a specific separator

In [92]:
word = ['Shivam', 'Sinha']

"-".join(word)

'Shivam-Sinha'

### ``replace()``

In [95]:
word = "Hello World"
print(word.replace('World', 'Duniya'))
print(word)

Hello Duniya
Hello World


> **Note:** ``replace()`` method returns a new string with the changes, **does not** modify the original string. (Strings are immutable)



### ``strip()``
It removes all leading and trailing whitespace (or specified characters) from a string.

In [98]:
"              Shivam       ".strip()

'Shivam'

---
# Practice Problem - String

### **Problem 1** : Find the length of a given string without using the ``len()`` function.

In [100]:
str = "Shivam Sinha"
length = 0

for i in str:
    length+=1

print("Length of the string is:", length)

Length of the string is: 12


### **Problem 2:** Extract username from a given E-mail.
Ex. if E-mail is ``shivamsinha@gmail.com``
then the username will be ``shivamsinha``

In [106]:
email = "shivamsinha@gmail.com"
temp_list = email.split('@')  # not string but a list
username = temp_list[0]
print("Username: ", username)

Username:  shivamsinha


"OR"


In [105]:
index = email.find('@')
#index = email.index('@')
username = email[:index]
print("Username: ", username)

Username:  shivamsinha


### **Problem 3:** WAP to count no. of occurance of a character in a given string.

In [111]:
str = "Shivam Sinha"
key = input("Enter: ")
freq = 0

for i in str:
    if i==key:
        freq+=1

print(key, "appears", freq, "times in the given string")

a appears 2 times in the given string


### **Problem 4:** WAP which can remove a particular character from a string.

In [116]:
str = "hello world"
key = input("Enter: ")
result_str = ""
for i in str:
    if i != key:
        result_str += i
    
print("Key =", key,"\nResult =", result_str)

Key = o 
Result = hell wrld


### **Problem 5:** WAP to check if a given string is palindrome or not

In [ ]:
# str = 'racecar'
str = 'shivam'

start = 0
end = len(str)-1

flag = True
while(start<end):
    if(str[start] == str[end]):
        start+=1
        end-=1
    else:
        flag = False
        break
    
if(flag):
    print("Yes, the given string is a palindrome")
else:
    print("The given string is not a palindrome")

The given string is not a palindrome


### **Program 6:** WAP to count no. of words in a string without using the ``split()`` function

In [129]:
str = "Shivam Sinha"
str = ""
str = "Hi, My name is Shivam Sinha"
count = 0

if(len(str)!=0):
    count+=1

for i in range(0,len(str)):
    if(str[i]==' '):
        count+=1

print(count)


6


"OR"

In [134]:
str = "Shivam Sinha"
list = []

temp = ''

for i in str:
    if(i != ' '):
        temp = temp + i
    else:
        list.append(temp)
        temp = ""
    
list.append(temp)
    
print(list)
print("No. of words in the given string is:", len(list) )

['Shivam', 'Sinha']
No. of words in the given string is: 2


### **Program 7:** WAP to convert a string to a title case without using the ``title()`` method

In [138]:
str = "hi how are you"

for i in str.split():
    print(i[0].upper() + i[1:].lower() , end = " ")

Hi How Are You 

### **Program 8:** WAP to convert an integer to a string



In [142]:
num = 456
digits = '0123456789'
result =''

while(num):
    result = digits[num%10] + result
    num = num//10

print(type(result))

<class 'str'>
